Application password experiment: Fall back to Jetpack proxy after app password generation fails

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

Closes WOOMOB-16079

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR updates error handling for application password generation by catching the failure to fall back to Jetpack proxy requests. Changes include:
- Updates  `RequestProcessorDelegate` to remove the reason in the delegate method. 
- Updates `RequestsProcessor` to only notify failure of application password generation upon catching errors directly related to application password.
- Updates `AlamofireNetwork` to catch `AFError.requestAdaptationFailed` and enable retrying requests when this error occurs.

## Testing steps
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Follow the instructions in TC3 and TC6 of the test plan pe5sF9-4Am-p2.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Tested the following test cases with simulator iPhone 16 iOS 18.2.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

No UI changes.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Attempt to fall back to Jetpack proxy after app password generation fails

itsmeichigo committed:
Do not notify failure for generic app password errors

itsmeichigo committed:
Fallback to Jetpack proxy when requestAdaptationFailed happens

itsmeichigo committed:
Update unit tests

itsmeichigo committed:
Fix failed unit test

RafaelKayumov reviewed (commented):
### For the TC3 I got:
```
{
  "error": "application_passwords_disabled",
  "message": "Application passwords are not available."
}
```
Requests fell back to the JP proxy.

### For the TC6 I got:
```
{
  "error": "general_error",
  "message": "Application passwords is not working."
}
```
The above error was returned couple times after every order lost loading attempt. I did 2 or 3 order list reloading by triggering pull-to-refresh. Then it crashed on the 3rd attempt:

<img width="1144" height="1010" alt="Снимок экрана 2025-09-07 в 23 02 39" src="https://github.com/user-attachments/assets/da8539c8-65c2-4861-9284-b6cfe805ba67" />

<img width="1391" height="1015" alt="Снимок экрана 2025-09-07 в 23 09 20" src="https://github.com/user-attachments/assets/5bcc70e9-dc45-48f2-b742-d631753d5190" />

Could be a race condition with a parallel `retriedJetpackRequests` object removing in separate threads. I'm able to reproduce by pull-to-refreshing dashboard and order list in parallel. Repro rate ~15%. Hard to reproduce.

itsmeichigo committed:
Move error handling to separate file and make it thread safe

itsmeichigo committed:
Merge branch 'trunk' into woomob-1126-retry-after-password-gen-fails

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo commented:
Great catch @RafaelKayumov! It's a noob error to update those array/dictionary for error handling in different threads. I moved all error handling to a separate file and use sync access to avoid race condition. I also added tests to confirm thread safety. Let me know if you have any other suggestions!

RafaelKayumov reviewed (approved):
Thx!
:shipit:
